// joint-plus/packages/joint-shapes-bpmn2/bpmn2/Activity.mjs
import { dia, util, V } from '@joint/core';
import { activityIcons, activityMarkers } from './icons.mjs';
import { borderSetAttributeWrapper, borderStyleSetAttribute, iconSetAttributeWrapper, IconsFlows, IconsOrigins, iconsPositionAttribute, iconsSetAttributeWrapper  } from './attributes.mjs';

export const Activity = dia.Element.define('bpmn2.Activity', {
    size: { width: 120, height: 100 },
    attrs: {
        root: {
            cursor: 'move',
            magnetSelector: 'background',
            highlighterSelector: 'background'
        },
        background: {
            refWidth: '100%',
            refHeight: '100%',
            fill: '#FFFFFF',
            rx: 10,
            ry: 10
        },
        border: {
            stroke: '#333333',
            fillRule: 'evenodd',
            borderType: 'single',
            borderStyle: 'solid',
            borderRadius: 10,
            strokeWidth: 2
        },
        icon: {
            iconColor: '#333333',
            iconType: 'none',
            x: 5,
            y: 0,
            width: 30,
            height: 30,
        },
        label: {
            refY: '50%',
            refX: '50%',
            textVerticalAnchor: 'middle',
            textAnchor: 'middle',
            fontSize: 12,
            fontFamily: 'sans-serif',
            fontWeight: 'bold',
            textWrap: {
                width: -10,
                height: -50,
                ellipsis: true
            }
        },
        markers: {
            event: 'element:marker:pointerdown',
            iconSize: 16,
            iconColor: '#333333',
            iconTypes: [''],
            iconsOrigin: IconsOrigins.bottomMiddle,
            iconsFlow: IconsFlows.row,
            refX: '50%',
            refY: '100%',
            refY2: -5
        }
    }
}, {
    markup: [{
        tagName: 'rect',
        selector: 'background'
    }, {
        tagName: 'image',
        selector: 'icon'
    }, {
        tagName: 'path',
        selector: 'border'
    }, {
        tagName: 'text',
        selector: 'label'
    }, {
        tagName: 'g',
        selector: 'markers'
    }]
}, {
    attributes: {
        'border-type': {
            set: borderSetAttributeWrapper((rect, attrs, offset) => {
                const borderRadius = attrs['border-radius'] ?? 0;
                const radius = Math.max(Math.min(borderRadius, 2), borderRadius - offset);
                const roundedRect = rect.toJSON();
                roundedRect.rx = roundedRect.ry = radius;
                return V.rectToPath(roundedRect);
            })
        },

        'border-style': {
            set: borderStyleSetAttribute
        },

        'icon-type': {
            set: iconSetAttributeWrapper('ACTIVITY_TYPE_ICONS')
        },

        'icon-types': {
            set: iconsSetAttributeWrapper('ACTIVITY_MARKER_ICONS'),
            position: iconsPositionAttribute
        },

        'icon-size': {
            // iconTypes attribute
        },
        'icons-origin': {
            // iconTypes attribute
        },
        'icons-flow': {
            // iconTypes attribute
        },
        'icon-color': {
            // iconType & iconTypes parameter
        },
        'border-radius': {
            // borderType parameter
        }
    },

    ACTIVITY_TYPE_ICONS: activityIcons,
    ACTIVITY_MARKER_ICONS: util.omit(activityMarkers, 'none')
});
