// joint-plus/packages/joint-color-palette/ColorPalette.mjs
import { SelectBox } from '../joint-select-box/SelectBox.mjs';
import { mvc } from '@joint/core';

const $ = mvc.$;

export const ColorPalette = SelectBox.extend({

    className: 'select-box color-palette',

    position: function() {

        var $selection = this.$('.select-box-selection');
        var selectionHeight = $selection.outerHeight();
        var selectionOffset = $selection.offset();

        var left = selectionOffset.left;
        var top = selectionOffset.top + selectionHeight;

        if (this.options.target !== document.body) {

            this.$target = this.$target || $(this.options.target);

            // Position relative to target element
            var targetOffset = this.$target.offset();
            left -= targetOffset.left - this.$target.scrollLeft();
            top -= targetOffset.top - this.$target.scrollTop();
        }

        this.$options.css({ left, top });
    }

}, {

    // Statics

    OptionsView: SelectBox.OptionsView.extend({

        renderOptionContent: function(option) {

            var $option = $('<div/>').addClass('select-box-option-content');

            $option.css('background-color', option.content);

            if (option.icon) {
                $option.prepend($('<img/>').addClass('select-box-option-icon').attr('src', option.icon));
            }

            return $option;
        }
    })

});
